### PR TITLE
Clean up the test code of `substring` kernel.

### LINKED ISSUE #1801: Clean up the testing code in `substring` kernel
Originally posted by @tustvold https://github.com/apache/arrow-rs/pull/1784#discussion_r890024346 and https://github.com/apache/arrow-rs/pull/1784#discussion_r890025561

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Closes #1801.

# Rationale for this change
Avoid repeated code.
More readable.
Less code is better.

# What changes are included in this PR?
Use macros.
Better format.

# Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by HaoYang670** [comment]
Blocked by #1852

**Comment by tustvold** [comment]
Sorry for the delay, I'll try to get to this later today if nobody else does, I'm currently on holiday and reviewing large PRs like this on a phone is challenging 😅

**Comment by HaoYang670** [comment]
Never mind. Have a good holiday! 🏝

File: arrow/src/compute/kernels/substring.rs [hunk 1]
```diff
@@ -445,64 +445,61 @@ mod tests {
...
+        ($input:expr, $(($start:expr, $len:expr, $result:expr)), *) => {
+            [
+                $(
+                    ($input.clone(), $start, $len, $result),
+                )*
+            ]
+        };
+    }
+
+    macro_rules! do_test {
```
Comment (alamb) [inline arrow/src/compute/kernels/substring.rs:458]: I think some docstrings explaining what was going on here (specifically what the expectations on `$cases) are would have helped me understand this code

File: arrow/src/compute/kernels/substring.rs [hunk 2]
```diff
@@ -445,64 +445,61 @@ mod tests {
...
+        ($cases:expr, $array_ty:ty, $substring_fn:ident) => {
+            $cases.into_iter().try_for_each::<_, Result<()>>(
```
Comment (alamb) [inline arrow/src/compute/kernels/substring.rs:460]: I wonder why bother with `try_for_each` and not just use `unwrap()` or `expect()`

Comment (HaoYang670) [reply arrow/src/compute/kernels/substring.rs:460]: Great catch! Updated.

File: arrow/src/compute/kernels/substring.rs (hunk 1 above)
Comment (HaoYang670) [reply arrow/src/compute/kernels/substring.rs:474]: Updated!

File: arrow/src/compute/kernels/substring.rs [hunk 3]
```diff
@@ -445,6 +445,16 @@ mod tests {
     use super::*;
     use crate::datatypes::*;
 
+    /// A helper macro to generate test cases.
```
Comment (alamb) [inline arrow/src/compute/kernels/substring.rs:448]: ❤️

**Review by alamb — APPROVED** [review]
Looks great -- thanks again @HaoYang670

**Pull request merged**